### DISCUSSION TITLE
How do I implement the equivalent of input() and print() in Kivy?

### DISCUSSION THREAD

**Discussion by EricTomaszewski** [kivy/kivy-github-discussions#6] [General]
Hi All.
Kivy seems to be a good GUI because it can be wrapped also into a mobile environment.
And I tested it with pyinstaller to make it into a simple .exe for "very non-techy clients".

Nevertheless, I'm trying to get a very BASIC functionality - namely:
an equivalent of input() => which seems to be TextInput()
and an equivalent of print()... here the story _seems_ a bit more complicated

Let's take "the simplest example of all" => this is myapp.py:
```
a = input("What's your starting number?")   # this would be replaced by TextInput
for i in range(0,10):
print(i)
time.sleep(i)
```

**How can I get this app running in my simple Kivy GUI environment below please?**
_(which has an input field, share button and output field that currently displays whatever is in the input field after pressing share button)_
Currently - integrating any input(), print() or time.sleep() causes the Kivy app to crash.

```
from kivy.app import App
from kivy.uix.boxlayout import BoxLayout
from kivy.uix.textinput import TextInput
from kivy.uix.label import Label
from kivy.uix.button import Button
from kivy.uix.scrollview import ScrollView
from kivy.uix.gridlayout import GridLayout
from kivy.clock import Clock
import time


class LinkShareApp(App):

    def build(self):
        self.title = 'Link Share App'
        layout = BoxLayout(orientation='vertical', padding=10, spacing=10)
        
        self.link_input = TextInput(hint_text="Enter the link to share")
        self.output_label = Label(text="Output will be displayed here")

        share_button = Button(text="Share Link")
        share_button.bind(on_press=self.share_link)

        layout.add_widget(self.link_input)
        layout.add_widget(share_button)
        layout.add_widget(self.output_label)
        
        return layout

    def share_link(self, instance):
        link = self.link_input.text
        self.output_label.text = f"Sharing link: {link}"

if __name__ == '__main__':
    LinkShareApp().run()

```

I checked the documentation and StackOverflow but could not find much but this post from 6 years ago which does not work when run in my VS Code Studio:
https://stackoverflow.com/questions/26265015/how-to-get-console-output-printed-using-kivy/26476526#26476526

**Can you point me in the right direction please?**

**Comment by kengoon** [comment]
Hi @EricTomaszewski , if my understanding is correct, you intend to link an additional terminal-based app to your existing GUI kivy app. Here are a few suggestions on the best approach:

1. Using `print` functions in kivy works well, but using `input` might pause the whole program until user input is provided, causing UI blockage. To avoid this, execute any UI-blocking tasks in a separate Thread to maintain the smooth flow of the UI.

2. If the terminal app is separate and you want to integrate it into kivy, consider using Kivy's oscpy. This enables you to establish a communication line between the kivy app and the terminal app using oscpy network protocols available at https://github.com/kivy/oscpy.

3. Alternatively, consider utilizing a separate process. Unlike a thread, it operates in a distinct process from the main app, which means direct communication isn't possible. However, there are workarounds for this, though my expertise in this area is limited.

**Reply by EricTomaszewski** [reply]
@kengoon thanks for this.

Ref 1:
I looked through the documentation: 
https://kivy.org/doc/stable/

but can't find anything that would present an example of:
 _To avoid this, execute any UI-blocking tasks in a separate Thread to maintain the smooth flow of the UI._

**Do you perhaps have any `code` reference in this area please?**

**Reply by kengoon** [reply]
@EricTomaszewski do you have any knowledge of thread?

**Reply by kengoon** [reply]
@EricTomaszewski I would recommend you read [this](https://realpython.com/intro-to-python-threading/) to grasp the basic knowledge of what threading is. You would figure out how to integrate it with Kivy once you have read it.

PS: I don't have any example code to show at the moment

**Comment by Julian-O** [comment]
@kengoon: I *think* Eric is asking for the output to appear within the app, not on an external terminal.

I assume that means writing to a Kivy label, rather than printing.

**Reply by EricTomaszewski** [reply]
@Julian-O indeed.

In simple terms. I'm trying to integrate this separate app.py:
```
a = input("What's your starting number?")   # this would be replaced by TextInput
for i in range(0,10):
print(i)
time.sleep(i)
```

into the main code to get variable i displayed **within** Kivy output window
but at the moment any tries to integrate & run this code causes Kivy GUI to crash.

I'm sure that there **must be** a way to integrate this simple code

- whether **internal** (within the main Kivy app window)
- or **external** (running it with `from app import name_of_the_function`)

**Reply by kengoon** [reply]
Hey @Julian-O , I understand your point. @EricTomaszewski , if we consider your input and print as representing TextInput and Label, there are two approaches to address this scenario:

1. Utilizing Kivy's clock to pause after displaying your text on the screen (assuming you're using Label and TextInput). However, using clock might result in a cumbersome solution due to significant callback management and asynchronous behavior, which could disrupt the loop.
2. Opting for [asynckivy](https://github.com/asyncgui/asynckivy) would be the more suitable solution. Here's a code snippet from its README:

```python
import asynckivy as ak

async def what_you_want_to_do(button):
    print('A')
    await ak.sleep(1)
    print('B')
    await ak.event(button, 'on_press')
    print('C')

ak.start(what_you_want_to_do(...))
```
Using the above, you can execute something like this:

```python
import asynckivy as ak

async def what_you_want_to_do(button):
    for i in range(1, 10):
        print(i)
        # Pause for 1 second without blocking the UI
        await ak.sleep(1) 

ak.start(what_you_want_to_do(...))
```
I believe the provided code is quite self-explanatory. If there's any confusion, feel free to ask."

**Reply by Julian-O** [reply]
I think we have some fundamental confusions here about what is being asked for. This isn't about threading.

The `print` statement, by default, sends output to `stdout`. When Kivy is running, it will continue to do that. Unless you have run with `pythonw.exe` (is that a Windows only concept? I am not sure) then the output will come out on the command line where you ran Kivy.

The OP doesn't want that to happen. They want to see the output in a Label.

The obvious way to do it is to replace `print(x)` with `my_label.text += str(x) + "\n"`

[I don't want to provide a full working example - perhaps someone else wants to.]

**Reply by kengoon** [reply]
I think otherwise, The OP made reference of `sleep`. In that case, maybe OP wants to wait before displaying the next text. For the `print` statement, I used it as an example to explain the context. I did not bother to take into account if the OP has the basic usage knowledge of Label or not. 
> Currently - integrating any input(), print() or time.sleep() causes the Kivy app to crash.

Judging from the above quote, the OP may really want to use `input` and `sleep` functions. Since `input` is UI blocking, I guess the app could be crashing at this point because it would cause a _Not_Responding_ error. Also I did not bother to take into account if the OP wants to use `TextInput` and `Label`.
>How can I get this app running in my simple Kivy GUI environment below please?

Here is another statement from the OP that may suggest that he wants to integrate `myapp.py`
>```
>a = input("What's your starting number?")   # this would be replaced by TextInput
>for i in range(0,10):
>print(I)
>time.sleep(i)
>```
in his kivy app. 

I think the OP owes us some further clarification of what he wants actually
